### ISSUE TITLE
saved_model.pb needs a different file extension

### ISSUE THREAD

**Issue by lutzroeder** [tensorflow/tensorflow#15061]
'saved_model.pb' file extension type clashes with every other '.pb' file created. Apps that register the '.pb' file extension won't be able to tell apart Saved Models from other protobuf files...

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by lutzroeder** [comment]
Have I written custom code: N/A
OS Platform and Distribution: N/A
TensorFlow installed from: pip
TensorFlow version: 1.4
Bazel version: N/A
CUDA/cuDNN version: N/A
GPU model and memory: N/A
Exact command to reproduce: N/A

**Comment by reedwm** [comment]
/CC @mrry can you comment? I'm not sure why the file extension is significant.

**Comment by mrry** [comment]
I'm also not sure why the file extension is significant. AIUI, `saved_model.pb` *is* a protobuf file, so an application that registers the `.pb` suffix in order to consume protobuf files ought to be able to consume a SavedModel.

@lutzroeder Can you describe what you're trying to do as a feature request, perhaps? Based on your post, I'm not convinced that the file format "needs" a different file extension, but if we understood more about what you're doing, that would help.

/cc @sukritiramesh @nfiedel

**Comment by lutzroeder** [comment]
If an app registers `.pb` it will list and open any protobuf file that uses the generic protobuf format, `saved_model.pb` or not. Since protobuf has no way of detecting the specific protobuf format (except for trying to decode the file) this can lead to some nasty conflicts. For example, an app registering '.pb' to open onnx files (which are protobuf as well) will now show `saved_model.pb` is a valid onnx file, except for it isn't. By analogy imagine all binary files using `.bin` instead of having `.doc`, `.xls` and `.ppt`. Word and Excel could just register `.bin` but wouldn't that be a recipe for disaster?

**Comment by mrry** [comment]
In that case, it sounds like the bug would be in the app that registers `*.pb` but only handles ONNX files.

Anyway, I have no skin in this game, so I'll let @nfiedel and/or @sukritiramesh comment on whether they'd consider a different extension for SavedModel files. I suspect backwards compatibility will be the main concern, since `saved_model.pb` is just one of the files in a standard directory layout that the tools use.

**Comment by nfiedel** [comment]
Thanks all for sharing context and the question here. For several reasons, I don't think it makes much sense to change the file extension of Saved Model's protobuffs.

First is forwards & backwards compatibility and migrations. There are 1000s of users of SavedModel. Changing the extension would require a long period (many months) of backwards compatibility along with forward-support, such that new code could read old models. We've done such migrations in the past and they involve a large amount of work and risk (i.e. risk of outages to customers using these models and code in production settings).

Second is a general question about file extensions and suitability of them for various purposes. There are a variety of generalized proto introspection tools out there, for which it could be perfectly reasonable to open any proto and view/inspect it by an extension ".pb". At the same time, relying on the file extension ".pb" of just one file in a ML model to specify the ML platform seems problematic.
 Remember that there are also weights and assets such as vocab or embedding files, that could be arbitrarily encoded as csv, proto, pbtxt, or otherwise.

In summary, I don't think it's particularly feasible to rename the saved_model.pb in the SavedModel specification, and even if we did, I don't think it will actually solve the problem at hand.

**Comment by lutzroeder** [comment]
I understand the the pain involved and why such changes aren't easy to make but also seems the like the current approach of everything is a `.pb` and there is no header to detect what is in the `.pb` isn't going to scale. Not expecting this to get fixed soon or easily but it sounds like the underlying issue of using `.pb` as a persistence and file exchange format needs some thinking.

I wrote a generic `.pb` browser and found it mostly useless as the files aren't very readable without the right `.proto` and root object being known. This leaves success of opening a `.pb` mostly to the user guessing correctly and/or having the right tribal knowledge being available on StackOverflow.

Currently tinkering with a [viewer for neural networks](https://www.github.com/lutzroeder/Netron) and seeing the reverse issue. It registers `.pb` and all sorts of files light up in Finder/Explorer classified as TensorFlow. Opening these files and having no way to tell if they contain `GraphDef` or `MetaGraphDef` or `SavedModel` or something else entirely leads to all kinds of asserts and crashes. For all the beauty of Protocol Buffers, this seems to be broken and does need fixing.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by reedwm** [comment]
Closing this issue since @nfiedel indicated `.pb` would not be changed.
